In [1]:
import numpy as np
from tensorflow import keras
from tensorflow.keras.datasets import imdb
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, LSTM, Bidirectional, Dense
from tensorflow.keras.preprocessing import sequence

In [2]:
max_features = 5000  # кількість слів, які будуть використовуватись в словнику
maxlen = 500  # максимальна довжина рецензії

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)

x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)

17464789/17464789 [==============================] - 1s 0us/step


Модель RNN

In [3]:
model = Sequential()
model.add(Embedding(max_features, 32))  # вбудовуємо слова у вектори
model.add(SimpleRNN(32))  # додаємо просту RNN
model.add(Dense(1, activation='sigmoid'))  # додаємо повнозв'язний шар з активацією sigmoid для класифікації

model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])

In [4]:
model.fit(x_train, y_train, batch_size=128, epochs=10, validation_data=(x_test, y_test))

Epoch 1/10


196/196 [==============================] - 17s 81ms/step - loss: 0.6338 - accuracy: 0.6244 - val_loss: 0.5178 - val_accuracy: 0.7493
Epoch 2/10
196/196 [==============================] - 16s 83ms/step - loss: 0.4380 - accuracy: 0.8083 - val_loss: 0.3940 - val_accuracy: 0.8337
Epoch 3/10
196/196 [==============================] - 16s 81ms/step - loss: 0.3478 - accuracy: 0.8581 - val_loss: 0.4208 - val_accuracy: 0.8290
Epoch 4/10
196/196 [==============================] - 16s 84ms/step - loss: 0.2957 - accuracy: 0.8787 - val_loss: 0.3851 - val_accuracy: 0.8469
Epoch 5/10
196/196 [==============================] - 18s 90ms/step - loss: 0.2402 - accuracy: 0.9054 - val_loss: 0.3870 - val_accuracy: 0.8458
Epoch 6/10
196/196 [==============================] - 18s 92ms/step - loss: 0.1956 - accuracy: 0.9242 - val_loss: 0.5184 - val_accuracy: 0.7772
Epoch 7/10
196/196 [==============================] - 19s 96ms/step - loss: 0.1531 - accuracy: 0.9423 - val_loss: 0.5019 - val_accurac

In [5]:
# Оцінка базової моделі RNN
scores_rnn = model.evaluate(x_test, y_test, verbose=0)
print("Base RNN Accuracy: %.2f%%" % (scores_rnn[1] * 100))

Base RNN Accuracy: 81.40%


Модель LSTM

In [6]:
model = Sequential()
model.add(Embedding(max_features, 32))  # вбудовуємо слова у вектори
model.add(LSTM(32))  # додаємо LSTM шар
model.add(Dense(1, activation='sigmoid'))  # додаємо повнозв'язний шар з активацією sigmoid для класифікації

model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])

model.fit(x_train, y_train, batch_size=128, epochs=10, validation_data=(x_test, y_test))

Epoch 1/10
196/196 [==============================] - 42s 205ms/step - loss: 0.5839 - accuracy: 0.6812 - val_loss: 0.6322 - val_accuracy: 0.6793
Epoch 2/10
196/196 [==============================] - 40s 204ms/step - loss: 0.3514 - accuracy: 0.8549 - val_loss: 0.3112 - val_accuracy: 0.8753
Epoch 3/10
196/196 [==============================] - 43s 219ms/step - loss: 0.2818 - accuracy: 0.8877 - val_loss: 0.3497 - val_accuracy: 0.8659
Epoch 4/10
196/196 [==============================] - 46s 237ms/step - loss: 0.2559 - accuracy: 0.9003 - val_loss: 0.3210 - val_accuracy: 0.8743
Epoch 5/10
196/196 [==============================] - 47s 242ms/step - loss: 0.2348 - accuracy: 0.9115 - val_loss: 0.3038 - val_accuracy: 0.8812
Epoch 6/10
196/196 [==============================] - 49s 252ms/step - loss: 0.2201 - accuracy: 0.9158 - val_loss: 0.2982 - val_accuracy: 0.8809
Epoch 7/10
196/196 [==============================] - 48s 243ms/step - loss: 0.2076 - accuracy: 0.9220 - val_loss: 0.2960 - val_ac

In [7]:
# Оцінка моделі з LSTM шарами
scores_lstm = model.evaluate(x_test, y_test, verbose=0)
print("LSTM Accuracy: %.2f%%" % (scores_lstm[1] * 100))

LSTM Accuracy: 87.33%


Двостороння модель

In [8]:
model = Sequential()
model.add(Embedding(max_features, 32))  # вбудовуємо слова у вектори
model.add(Bidirectional(SimpleRNN(32)))  # додаємо двосторонню мережу з простою RNN
model.add(Dense(1, activation='sigmoid'))  # додаємо повнозв'язний шар з активацією sigmoid для класифікації

model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])

model.fit(x_train, y_train, batch_size=128, epochs=10, validation_data=(x_test, y_test))

Epoch 1/10
196/196 [==============================] - 24s 114ms/step - loss: 0.5888 - accuracy: 0.6804 - val_loss: 0.4816 - val_accuracy: 0.7802
Epoch 2/10
196/196 [==============================] - 28s 142ms/step - loss: 0.4013 - accuracy: 0.8301 - val_loss: 0.4180 - val_accuracy: 0.8170
Epoch 3/10
196/196 [==============================] - 23s 115ms/step - loss: 0.3427 - accuracy: 0.8582 - val_loss: 0.4024 - val_accuracy: 0.8194
Epoch 4/10
196/196 [==============================] - 23s 116ms/step - loss: 0.3104 - accuracy: 0.8737 - val_loss: 0.4257 - val_accuracy: 0.8020
Epoch 5/10
196/196 [==============================] - 23s 115ms/step - loss: 0.2944 - accuracy: 0.8838 - val_loss: 0.8935 - val_accuracy: 0.7011
Epoch 6/10
196/196 [==============================] - 23s 116ms/step - loss: 0.2787 - accuracy: 0.8924 - val_loss: 0.3727 - val_accuracy: 0.8430
Epoch 7/10
196/196 [==============================] - 23s 118ms/step - loss: 0.2392 - accuracy: 0.9072 - val_loss: 0.3699 - val_ac

In [9]:
# Оцінка моделі з Bidirectional RNN
scores_lstm = model.evaluate(x_test, y_test, verbose=0)
print("Bidirectional RNN Accuracy: %.2f%%" % (scores_lstm[1] * 100))

Bidirectional RNN Accuracy: 86.02%


Глибока модель

In [10]:
model = Sequential()
model.add(Embedding(max_features, 32))  # вбудовуємо слова у вектори
model.add(LSTM(32, return_sequences=True))  # додаємо перший LSTM шар з поверненням послідовностей
model.add(LSTM(32))  # додаємо другий LSTM шар
model.add(Dense(1, activation='sigmoid'))  # додаємо повнозв'язний шар з активацією sigmoid для класифікації

model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])

model.fit(x_train, y_train, batch_size=128, epochs=10, validation_data=(x_test, y_test))

Epoch 1/10
196/196 [==============================] - 114s 571ms/step - loss: 0.5403 - accuracy: 0.7034 - val_loss: 0.4019 - val_accuracy: 0.8376
Epoch 2/10
196/196 [==============================] - 113s 579ms/step - loss: 0.3335 - accuracy: 0.8627 - val_loss: 0.3491 - val_accuracy: 0.8482
Epoch 3/10
196/196 [==============================] - 112s 572ms/step - loss: 0.2766 - accuracy: 0.8911 - val_loss: 0.3048 - val_accuracy: 0.8778
Epoch 4/10
196/196 [==============================] - 113s 578ms/step - loss: 0.2497 - accuracy: 0.9028 - val_loss: 0.2928 - val_accuracy: 0.8779
Epoch 5/10
196/196 [==============================] - 111s 569ms/step - loss: 0.2287 - accuracy: 0.9108 - val_loss: 0.2872 - val_accuracy: 0.8810
Epoch 6/10
196/196 [==============================] - 112s 570ms/step - loss: 0.2083 - accuracy: 0.9218 - val_loss: 0.3065 - val_accuracy: 0.8715
Epoch 7/10
196/196 [==============================] - 98s 500ms/step - loss: 0.1972 - accuracy: 0.9247 - val_loss: 0.3310 - 

In [11]:
# Оцінка моделі з глибокою мережею LSTM
scores_deep_lstm = model.evaluate(x_test, y_test, verbose=0)
print("Deep LSTM Accuracy: %.2f%%" % (scores_deep_lstm[1] * 100))

Deep LSTM Accuracy: 86.56%


Результати:
- Base RNN Accuracy: 81.40%
- LSTM Accuracy: 87.33%
- Bidirectional RNN Accuracy: 86.02%
- Deep LSTM Accuracy: 86.56%

В цілому результати точності (за винятком Base RNN) схожі, але все ж для данного випадку кращим вибором є LSTM модель, адже вона була доволі швидкою при максимальному результату точності. 